# PRÀCTICA BLOC II (INTEL·LIGÈNCIA ARTIFICIAL) 



Miquel Vives Marcus<br />
Repositori Git: https://github.com/miqvm/practica-ML-21722

## Importam les llibreries

Per resoldre aquesta pràctica utilitzarem la llibraria sklearn, per realitzar tot el relacionat amb el machine learning; pandas per llegir i operar sobre el dataset i numpy per operar amb arrays i matrius. 

S'importen les llibraries esmentades:

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

## Llegim les dades

Per fer-ho, usarem ``pandas`` i una utilitat de la llibraria ``os``

In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]


## Analisis de les dades

En aquest bloc, s'identificaran els diferents atributs que tenen els registres del dataset, s'eliminaran els irrelevants, i es convertiran a un format apte per treballar els que no ho estiguin.

Primerament, mostrarem el dataset per complet:

In [3]:
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

**Descripció dels atributs:**

``Unnamed 0:`` Posició del jugador en el dataset.<br />
``ID:`` Nombre identificador del jugador en la base de dades de FIFA.<br />
``Name:`` Nom del jugador.<br />
``Age:`` Edat del jugador.<br />
``Photo:`` Indica l'URL de la foto del jugador.<br />
``Nationality:`` País de naixament.<br />
``Flag:`` URL de la bandera del país de naixament.<br />
``Overall`` Mitja dels seus atributs, a major és el nombre, millor és el jugador.<br />
``Potential:`` Indica el potencial del jugador.<br />
``Special:`` Indica si el jugador té una carta especial<br />
``Club:`` Equip al que pertany.<br />
``Club Logo:`` URL de l'emblema del club.<br />
``Value:`` Preu de mercat, atribut a fer la predicció.<br />
``Wage:`` Salari del jugador.<br />
``Special Preferred Foot:`` Indica si es dretà o esquerrà.<br />
``International Reputation:`` De 0-5 indica quan de valorat és aquest jugador al reu del món.<br />
``Weak Foot:`` De 0-5 indica quan és la seva destresa amb el peu dolent.<br />
``Skill Moves:`` De 0-5 indica l'habilitat que té per fer filigranes.<br />
``Work Rate:`` Mesurat amb Low/Medium/High indica el ritme de treball.<br />
``Body Type:`` Model del cos 3D en el FIFA.<br />
``Real Face:`` Indica si la cara que apareix en el joc es la seva real.<br />
``Position:`` Indica la seva posició en el camp de joc.<br />
``Jersey Number:`` Nombre que duu a la samarreta en el seu equip.<br />
``Joined:`` Quan de temps fa que aquest jugador està en el club que s'indica.<br />
``Loaned From:`` Indica si el jugador ha estat cedit per un altre club.<br />
``Contract Valid Until:`` Data d'expiració del contracte.<br />
``Height:`` Altura del jugador.<br />
``Weight:`` Pes del jugador.<br />
``LS:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left striker".<br />
``ST:`` Indica de 0-100 l'habilitat del jugador en la posició de "Striker".<br />
``RS:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right striker".<br />
``LW:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left winger".<br />
``LF:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left forward".<br />
``CF:`` Indica de 0-100 l'habilitat del jugador en la posició de "Central forward".<br />
``RF:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right forward".<br />
``RW:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right winger".<br />
``LAM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left attacking midfielder".<br />
``CAM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Central attacking midfielder".<br />
``RAM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right attacking midfielder".<br />
``LM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left midfielder".<br />
``LCM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left central midfielder".<br />
``CM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Central midfielder".<br />
``RCM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right central midfielder".<br />
``RM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right midfielder".<br />
``LWB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left winger back".<br />
``LDM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left defense midfielder".<br />
``CDM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Central defense midfielder".<br />
``RDM:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right defense midfielder".<br />
``RWB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right winger back".<br />
``LB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left back".<br />
``LCB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Left central back".<br />
``CB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Central back".<br />
``RCB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right central back".<br />
``RB:`` Indica de 0-100 l'habilitat del jugador en la posició de "Right back".<br />
``Crossing:``  Indica de 0-100 la presició del jugador en fer passades creuades.<br />
``Finishing:`` Indica de 0-100 la presició del jugador en fer xuts dins l'àrea.<br />
``HeadingAccuracy:`` Indica de 0-100 l'habilitat del jugador en fer remats de cap.<br />
``ShortPassing:`` Indica de 0-100 l'habilitat del jugador en fer passades a curta distància.<br />
``Volleys Dribbling:`` Indica de 0-100 l'habilitat del jugador en ragetetjar usant salvaments.<br />
``Curve:`` Indica de 0-100 l'habilitat del jugador en fer un tir amb curvatura.<br />
``FKAccuracy:``  Indica de 0-100 l'habilitat del jugador per xutar un tir lliure.<br />
``LongPassing:`` Indica de 0-100 l'habilitat del jugador en fer passades a llarga distància.<br />
``BallControl:`` Indica de 0-100 l'habilitat del jugador en controlar la pilota.<br />
``Acceleration:`` Indica de 0-100 l'habilitat del jugador en accelerar.<br />
``SprintSpeed:`` Indica de 0-100 la velocitat del jugador.<br />
``Agility Reactions:`` Indica de 0-100 l'agiliat del jugador.<br />
``Balance:`` Indica de 0-100 l'habilitat del jugador en mantenir l'equilibri després d'un cop.<br />
``ShotPower:`` Indica de 0-100 la potència de xut del jugador.<br />
``Jumping Stamina Strength:`` Indica de 0-100 l'altura que pot botar el jugador.<br />
``Long Shots:`` Indica de 0-100 la presició del jugador en xuts a llarga distància.<br />
``Aggression:`` Indica de 0-100 l'habilitat del jugador en realitzar una falta.<br />
``Interceptions:`` Indica de 0-100 l'habilitat del jugador en rebre i mantenir la pilota.<br />
``Positioning:`` Indica de 0-100 l'habilitat del jugador en posicionar-se en el camp de joc.<br />
``Vision:`` Indica de 0-100 l'habilitat del jugador en percebre als seus companys i l'entorn de joc durant el partit.<br />
``Penalties:`` Indica de 0-100 l'habilitat del jugador en xutar penaltis.<br />
``Composure:`` Indica de 0-100 l'habilitat del jugador en aguantar la presió frent als rivals.<br />
``Marking Standing Tackle:`` Indica de 0-100 l'habilitat del jugador en fer presió a un jugador rival.<br />
``SlidingTackle:`` Indica de 0-100 l'habilitat del jugador en realitzar una falta a un jugador rival.<br />
``GKDiving:`` De 0-100, habilitat del porter per agafar la pilota mentres ell està a l'aire.<br />
``GKHandling:`` De 0-100, habilitat del porter per aturar la pilota i mantenir-la en les seves mans.<br />
``GKKicking:`` De 0-100, habilitat del porter per xutar la pilota.<br />
``GKPositioning:`` De 0-100, habilitat del porter per recolocar-se una vegada aturi el primer xut a porta.<br />
``GKReflexes:`` De 0-100, reflexes del porter.<br />
``Release Clause:`` Quantitat monetaria per realitzar una clausula de recessió.<br />

In [4]:
df.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number      Crossing     Finishing  HeadingAccuracy  \
count   18147.000000  18159.000000  18159.000000     18159.000000   
mean       19.546096     49.734181     45.550911        52.298144   
std        15.947765     18.364524     19.525820        17.379909   
min         1.000000      5.000000      2.000000         4.000000   
25%         8.000000     38.000000     30.000000        44.000000   
50%        17.000000     54.000000     49.000000        56.000000   
75%        26.000000     64.000000     62.000000        64.000000   
max        99.000000     93.000000     95.000000        94.000000   

       ShortPassing       Volleys     Dribbling         Curve    FKAccuracy  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      58.686712     42.909026     55.371001     47.170824     42.863153   
std       14.699495     17.694408     18.910371     18.395264     17.478763   
min        7.000000      4.000000      4.000000      6.000000      3.000000   
25%       54.000000     30.000000     49.000000     34.000000     31.000000   
50%       62.000000     44.000000     61.000000     48.000000     41.000000   
75%       68.000000     57.000000     68.000000     62.000000     57.000000   
max       93.000000     90.000000     97.000000     94.000000     94.000000   

        LongPassing   BallControl  Acceleration   SprintSpeed       Agility  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      52.711933     58.369459     64.614076     64.726967     63.503607   
std       15.327870     16.686595     14.927780     14.649953     14.766049   
min        9.000000      5.000000     12.000000     12.000000     14.000000   
25%       43.000000     54.000000     57.000000     57.000000     55.000000   
50%       56.000000     63.000000     67.000000     67.000000     66.000000   
75%       64.000000     69.000000     75.000000     75.000000     74.000000   
max       93.000000     96.000000     97.000000     96.000000     96.000000   

          Reactions       Balance     ShotPower       Jumping       Stamina  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      61.836610     63.966573     55.460047     65.089432     63.219946   
std        9.010464     14.136166     17.237958     11.820044     15.894741   
min       21.000000     16.000000      2.000000     15.000000     12.000000   
25%       56.000000     56.000000     45.000000     58.000000     56.000000   
50%       62.000000     66.000000     59.000000     66.000000     66.00000


Amb la descripció feta dels atributs, s'eliminaran els que no aporten res a la predicció, aquests són:

``Name:`` Ja que el nom del jugador com a tal no afecta al preu d'ell en el mercat.<br />
``Photo:`` URL d'una imatge.<br />
``Flag:`` URL d'una imatge.<br />
``Club Logo:`` URL d'una imatge.<br />
``Body Type:`` El model de cos 3D en el joc no aporta res a la predicció.<br />
``Real Face:`` Si té o no cara en el FIFA no aporta res a la predicció.<br />
``Joined:`` Baix el meu punt de vista no afecta al preu del jugador la quantitat de temps que duu jugant a un mateix club, ja que el que importa són els mèrtits personals no la durabilitat en el club. Pot pertanyer a un club durant molt de temps però no aportar res a aquest.<br />
``Loaned From:`` Baix el meu punt de vista no afecta al preu del jugador per quin equip està cedit, ja que tendrà més importància saber el club al que pertany pel prestigui que pugui obtenir d'ell.<br />
``Unnamed: 0:`` El nombre d'identificació en el dataset no aporta informació per la predicció.<br />
``ID:`` El nombre d'identificació en el FIFA no aporta informació en la predicció.<br />
``Special:`` Si té o no cartes especials no aporta informació en la predicció.<br />
``Jersey Number:`` El nombre de dorsal no aporta informació a la predicció.<br />

In [5]:
df = df.drop(['Unnamed: 0', 'ID', 'Special', 'Jersey Number', 'Name', 'Photo', 'Flag', 'Club Logo', 'Body Type', 'Real Face', 'Joined', 'Loaned From'], axis=1)
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

        Work Rate Position Contract Valid Until Height  Weight    LS    ST  \
0  Medium/ Medium       RF                 2021    5'7  159lbs  88+2  88+2   
1       High/ Low       ST                 2022    6'2  183lbs  91+3  91+3   
2    High/ Medium       LW                 2022    5'9  150lbs  84+3  84+3   
3  Medium/ Medium       GK                 2020    6'4  168lbs   NaN   NaN   
4      High/ High      RCM                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0        

### Eliminació de NaNS

Els NaNs ens impediran realitzar estimacións correctes, per tant es important no tenir-ne alhora de realitzar les prediccions. Per saber quines columnes tenen NaNs realitzarem la seguent instrucció:

In [6]:
df.columns[df.isna().any()].tolist()

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

In [7]:
df[df.isna().any(axis=1)]


Age Nationality  Overall  Potential               Club   Value   Wage  \
3       27       Spain       91         93  Manchester United    €72M  €260K   
9       25    Slovenia       90         93    Atlético Madrid    €68M   €94K   
18      26     Germany       89         92       FC Barcelona    €58M  €240K   
19      26     Belgium       89         90        Real Madrid  €53.5M  €240K   
22      32     Germany       89         89  FC Bayern München    €38M  €130K   
...    ...         ...      ...        ...                ...     ...    ...   
18178   18      Sweden       48         65         Dalkurd FF    €50K    €1K   
18180   22    Scotland       48         58   St. Johnstone FC    €40K    €1K   
18183   44     England       48         48   Cambridge United      €0    €1K   
18194   18       Italy       47         65              Lecce    €50K    €1K   
18198   18     England       47         70      Burton Albion    €60K    €1K   

      Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
3              Right                       4.0        3.0          1.0   
9              Right                       3.0        3.0          1.0   
18             Right                       3.0        4.0          1.0   
19              Left                       4.0        2.0          1.0   
22             Right                       5.0        4.0          1.0   
...              ...                       ...        ...          ...   
18178          Right                       1.0        2.0          1.0   
18180          Right                       1.0        2.0          1.0   
18183          Right                       1.0        2.0          1.0   
18194          Right                       1.0        3.0          1.0   
18198          Right                       1.0        2.0          1.0   

            Work Rate Position Contract Valid Until Height  Weight   LS   ST  \
3      Medium/ Medium       GK                 2020    6'4  168lbs  NaN  NaN   
9      Medium/ Medium       GK                 2021    6'2  192lbs  NaN  NaN   
18     Medium/ Medium       GK                 2022    6'2  187lbs  NaN  NaN   
19     Medium/ Medium       GK                 2024    6'6  212lbs  NaN  NaN   
22     Medium/ Medium       GK                 2021    6'4  203lbs  NaN  NaN   
...               ...      ...                  ...    ...     ...  ...  ...   
18178  Medium/ Medium       GK                 2018    6'0  176lbs  NaN  NaN   
18180  Medium/ Medium       GK                 2019    6'1  172lbs  NaN  NaN   
18183  Medium/ Medium       GK                 2023    6'2  190lbs  NaN  NaN   
18194  Medium/ Medium       GK                 2021    6'3  187lbs  NaN  NaN   
18198  Medium/ Medium       GK                 2021   5'11  154lbs  NaN  NaN   

        RS   LW   LF   CF   RF   RW  LAM  CAM  RAM   LM  LCM   CM  RCM   RM  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
18178  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18180  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18183  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18194  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18198  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       LWB  LDM  CDM  RDM  RWB   LB  LCB   CB  RCB   RB  Crossing  Finishing  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      17.0       13.0   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      13.0       11.0   
18     

Podem veure com els porters no tenen cap valor als atributs de posició (*LS, ST, RS, LW, LF, CF, RF, RW, LAM, CAM, RAM, LM, LCM, CM, RCM, RM, LWB, LDM, CDM, RDM, RWB, LB, LCB, CB, RCB i RB*) per tant al ser un valor suficientment relevant per realitzar la predicció del preu d'un jugador, s'hauran d'assignar algun valor, ja que si s'eliminessin, no es podria determinar el preu dels porters correctament, per tant canviarem aquest valor NaN per un 0.

In [8]:
df["LS"] = df["LS"].fillna("0+0")
df["ST"] = df["ST"].fillna("0+0")
df["RS"] = df["RS"].fillna("0+0")
df["LW"] = df["LW"].fillna("0+0")
df["LF"] = df["LF"].fillna("0+0")
df["CF"] = df["CF"].fillna("0+0")
df["RF"] = df["RF"].fillna("0+0")
df["RW"] = df["RW"].fillna("0+0")
df["LAM"] = df["LAM"].fillna("0+0")
df["CAM"] = df["CAM"].fillna("0+0")
df["RAM"] = df["RAM"].fillna("0+0")
df["LM"] = df["LM"].fillna("0+0")
df["LCM"] = df["LCM"].fillna("0+0")
df["CM"] = df["CM"].fillna("0+0")
df["RCM"] = df["RCM"].fillna("0+0")
df["RM"] = df["RM"].fillna("0+0")
df["LWB"] = df["LWB"].fillna("0+0")
df["LDM"] = df["LDM"].fillna("0+0")
df["CDM"] = df["CDM"].fillna("0+0")
df["RDM"] = df["RDM"].fillna("0+0")
df["RWB"] = df["RWB"].fillna("0+0")
df["LB"] = df["LB"].fillna("0+0")
df["LCB"] = df["LCB"].fillna("0+0")
df["CB"] = df["CB"].fillna("0+0")
df["RCB"] = df["RCB"].fillna("0+0")
df["RB"] = df["RB"].fillna("0+0")

df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

        Work Rate Position Contract Valid Until Height  Weight    LS    ST  \
0  Medium/ Medium       RF                 2021    5'7  159lbs  88+2  88+2   
1       High/ Low       ST                 2022    6'2  183lbs  91+3  91+3   
2    High/ Medium       LW                 2022    5'9  150lbs  84+3  84+3   
3  Medium/ Medium       GK                 2020    6'4  168lbs   0+0   0+0   
4      High/ High      RCM                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0        

Tornarem a mirar quines columnes tenim amb NaNs

In [9]:
df[df.isna().any(axis=1)]


Age Nationality  Overall  Potential                            Club  \
28      26    Colombia       88         89               FC Bayern München   
38      30   Argentina       88         88                           Milan   
91      29      Brazil       85         85  Guangzhou Evergrande Taobao FC   
166     24      Brazil       83         90  Guangzhou Evergrande Taobao FC   
176     24     Croatia       83         89                         Chelsea   
...    ...         ...      ...        ...                             ...   
17979   21    China PR       51         60             Guizhou Hengfeng FC   
18026   21    China PR       50         59             Guizhou Hengfeng FC   
18031   20    China PR       50         61                  Stabæk Fotball   
18056   19       Italy       50         65                          Ascoli   
18183   44     England       48         48                Cambridge United   

        Value   Wage Preferred Foot  International Reputation  Weak Foot  \
28     €69.5M  €315K           Left                       4.0        3.0   
38       €57M  €245K          Right                       4.0        4.0   
91       €37M  €235K          Right                       3.0        3.0   
166    €36.5M   €18K           Left                       2.0        4.0   
176      €35M  €165K          Right                       3.0        4.0   
...       ...    ...            ...                       ...        ...   
17979    €60K    €2K          Right                       1.0        2.0   
18026    €50K    €2K          Right                       1.0        2.0   
18031    €40K    €2K          Right                       1.0        3.0   
18056    €60K    €3K           Left                       1.0        3.0   
18183      €0    €1K          Right                       1.0        2.0   

       Skill Moves       Work Rate Position Contract Valid Until Height  \
28             4.0  Medium/ Medium      LAM         Jun 30, 2019   5'11   
38             3.0    High/ Medium       LS         Jun 30, 2019    6'1   
91             3.0      High/ High      LDM         Jun 30, 2019    6'0   
166            4.0    High/ Medium      CAM         Dec 31, 2018    6'3   
176            4.0      High/ High      LCM         Jun 30, 2019    5'9   
...            ...             ...      ...                  ...    ...   
17979          2.0  Medium/ Medium       CM         Dec 31, 2018    6'0   
18026          2.0  Medium/ Medium       LM         Dec 31, 2018   5'10   
18031          2.0  Medium/ Medium       RB         Dec 31, 2018    5'7   
18056          2.0  Medium/ Medium       CM         Jun 30, 2019   5'10   
18183          1.0  Medium/ Medium       GK                 2023    6'2   

       Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
28     172lbs  80+3  80+3  80+3  84+3  83+3  83+3  83+3  84+3  85+3  85+3   
38     196lbs  85+3  85+3  85+3  81+3  83+3  83+3  83+3  81+3  80+3  80+3   
91     157lbs  81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3  81+3  81+3   
166    176lbs  80+2  80+2  80+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2   
176    172lbs  71+3  71+3  71+3  79+3  78+3  78+3  78+3  79+3  81+3  81+3   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17979  172lbs  49+2  49+2  49+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2   
18026  143lbs  44+2  44+2  44+2  47+2  46+2  46+2  46+2  47+2  48+2  48+2   
18031  143lbs  37+2  37+2  37+2  42+2  40+2  40+2  40+2  42+2  39+2  39+2   
18056  143lbs  44+2  44+2  44+2  47+2  45+2  45+2  45+2  47+2  47+2  47+2   
18183  190lbs   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
28     85+3  83+3  81+3  81+3  81+3  83+3  69+3  68+3  68+3  68+3  69+3  65+3   
38     80+3  78+3  71+3  71+3  71+3  78+3  55+3  53+3  53+3  53+3  55+3  51+3   
91     81+3  79+3  82+3  82+3  82+3  79+3  78+3  81+3  81+3  81+3  78+3  78+3   
166    

In [10]:
df.columns[df.isna().any()].tolist()

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

El club al que pertanyen pot causar una gran diferencia al valor d'un jugador per causes com pugui ser el prestigui que aporta la pertenyença per tant també s'eliminaran aquests jugadors. 

In [11]:
df = df.dropna(subset = ['Club'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

        Work Rate Position Contract Valid Until Height  Weight    LS    ST  \
0  Medium/ Medium       RF                 2021    5'7  159lbs  88+2  88+2   
1       High/ Low       ST                 2022    6'2  183lbs  91+3  91+3   
2    High/ Medium       LW                 2022    5'9  150lbs  84+3  84+3   
3  Medium/ Medium       GK                 2020    6'4  168lbs   0+0   0+0   
4      High/ High      RCM                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0        

In [12]:
df.columns[df.isna().any()].tolist()

['Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

Veim que l'atribut "Release Clause" amb valors NaNs, es suposarà que si el jugador no té cap clausula de recesió aquesta serà 0 per tant:

In [13]:
df["Release Clause"] = df["Release Clause"].fillna("€0")

Es miren de nou els valors NaN:

In [14]:
df.columns[df.isna().any()].tolist()

['Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Contract Valid Until',
 'Height',
 'Weight',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes']

In [15]:
df[df.isna().any(axis=1)]

Age       Nationality  Overall  Potential  \
13236   33          Scotland       62         62   
13237   29         Nicaragua       62         62   
13238   35           England       62         62   
13239   20             Italy       62         72   
13240   24           England       62         66   
13241   21            Poland       62         72   
13242   23         Argentina       62         70   
13243   19           England       62         78   
13244   30           Denmark       62         62   
13245   21            Brazil       62         69   
13246   22            France       62         68   
13247   28           Austria       62         62   
13248   24  Northern Ireland       62         69   
13249   27          China PR       62         62   
13250   29          Paraguay       62         62   
13251   34           Belgium       62         62   
13252   22           Albania       62         70   
13253   31          Scotland       62         62   
13254   17           Germany       62         82   
13255   25           Germany       62         65   
13256   26           England       62         62   
13257   25    Korea Republic       62         65   
13258   23      Saudi Arabia       62         67   
13259   27             Chile       62         65   
13260   20             Spain       62         69   
13261   25      Saudi Arabia       62         64   
13262   24           Germany       62         66   
13263   25           Austria       62         66   
13264   29            France       62         62   
13265   28           England       62         62   
13266   20            Sweden       62         73   
13267   22          Colombia       62         70   
13268   37            Poland       62         62   
13269   26          Colombia       62         64   
13270   19            Brazil       62         77   
13271   26             Chile       62         63   
13272   22          Scotland       62         70   
13273   25           Estonia       62         67   
13274   27             Italy       62         62   
13275   23         Venezuela       62         70   
13276   26      Saudi Arabia       62         63   
13277   26             Chile       62         62   
13278   17       Netherlands       62         79   
13279   22             Italy       62         69   
13280   19            France       62         77   
13281   27    Korea Republic       62         62   
13282   25            Mexico       62         65   
13283   25          China PR       62         66   

                                Club  Value Wage Preferred Foot  \
13236                       Rochdale  €120K  €1K            NaN   
13237                Boyacá Chicó FC  €300K  €1K            NaN   
13238                   Notts County  €140K  €3K            NaN   
13239                        Brescia  €425K  €1K            NaN   
13240         Hamilton Academical FC  €400K  €1K            NaN   
13241                  Śląsk Wrocław  €425K  €1K            NaN   
13242         Club Atlético Aldosivi  €450K  €2K            NaN   
13243                        Everton  €600K  €5K            NaN   
13244                       Hobro IK  €230K  €2K            NaN   
13245                   HJK Helsinki  €425K  €1K            NaN   
13246                     AS Béziers  €425K  €2K            NaN   
13247                 SV Mattersburg  €240K  €3K            NaN   
13248                Tranmere Rovers  €375K  €2K            NaN   
13249  Shanghai Greenland Shenhua FC  €250K  €3K            NaN   
13250               Itagüí Leones FC  €300K  €1K            NaN   
13251                      NAC Breda  €150K  €2K            NaN   
13252                       Malmö FF  €375K  €1K            NaN   
13253                Carlisle United  €200K  €2K            NaN   
13254                  VfB Stuttgart  €550K  €2K            NaN   
13255         Hamilton Academical FC  €325K  €1K            NaN   
13256                      Dundee FC  €325K  €1K   

Es pot veure com hi ha una serie de jugadors que tenen una gran quantitat d'atributs fora assignar, per tant s'eliminaran aquests jugadors ja que no aporten informació útil per realitzar la predicció. Al borrar tots els jugadors que tenen Preferred Foot = NaN, ja entren implicitament tots els altres altres atributs = NaN ja que són els unics jugadors que ho tenen així, per tant basta eliminar els jugadors que tenen Preferred Foot fora assignar.

In [16]:
df = df.dropna(subset = ['Preferred Foot'])
df.columns[df.isna().any()].tolist()

[]

Es pot veure com ja no tenim cap NaN a la taula, per tant podem proseguir.

### Conversió d'Strings a format utilitzable

L'algoritme de Machine Learning no sap tractar els atributs en format String, per tant aquests s'hauran de transformar a algun format que l'algoritme ho pugui entendre, per això usarem la funció get_dummies proporcionada per la llibreria ``panda``.

Per usar el Club al que pertany un jugador per determinar el seu preu de mercat, al trobar-se en format String s'haurà de transformar:

In [17]:
clb = df.pop("Club")

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)
df

Age Nationality  Overall  Potential    Value   Wage Preferred Foot  \
0       31   Argentina       94         94  €110.5M  €565K           Left   
1       33    Portugal       94         94     €77M  €405K          Right   
2       26      Brazil       92         93  €118.5M  €290K          Right   
3       27       Spain       91         93     €72M  €260K          Right   
4       27     Belgium       91         92    €102M  €355K          Right   
...    ...         ...      ...        ...      ...    ...            ...   
17913   19     England       47         65     €60K    €1K          Right   
17914   19      Sweden       47         63     €60K    €1K          Right   
17915   16     England       47         67     €60K    €1K          Right   
17916   17     England       47         66     €60K    €1K          Right   
17917   16     England       46         66     €60K    €1K          Right   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
0                           5.0        4.0          4.0  Medium/ Medium   
1                           5.0        4.0          5.0       High/ Low   
2                           5.0        5.0          5.0    High/ Medium   
3                           4.0        3.0          1.0  Medium/ Medium   
4                           4.0        5.0          4.0      High/ High   
...                         ...        ...          ...             ...   
17913                       1.0        2.0          2.0  Medium/ Medium   
17914                       1.0        2.0          2.0  Medium/ Medium   
17915                       1.0        3.0          2.0  Medium/ Medium   
17916                       1.0        3.0          2.0  Medium/ Medium   
17917                       1.0        3.0          2.0  Medium/ Medium   

      Position Contract Valid Until Height  Weight    LS    ST    RS    LW  \
0           RF                 2021    5'7  159lbs  88+2  88+2  88+2  92+2   
1           ST                 2022    6'2  183lbs  91+3  91+3  91+3  89+3   
2           LW                 2022    5'9  150lbs  84+3  84+3  84+3  89+3   
3           GK                 2020    6'4  168lbs   0+0   0+0   0+0   0+0   
4          RCM                 2023   5'11  154lbs  82+3  82+3  82+3  87+3   
...        ...                  ...    ...     ...   ...   ...   ...   ...   
17913       CM                 2019    5'9  134lbs  42+2  42+2  42+2  44+2   
17914       ST                 2020    6'3  170lbs  45+2  45+2  45+2  39+2   
17915       ST                 2021    5'8  148lbs  45+2  45+2  45+2  45+2   
17916       RW                 2019   5'10  154lbs  47+2  47+2  47+2  47+2   
17917       CM                 2019   5'10  176lbs  43+2  43+2  43+2  45+2   

         LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0      93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1      90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2      89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3       0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4      87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17913  44+2  44+2  44+2  44+2  45+2  45+2  45+2  44+2  45+2  45+2  45+2  44+2   
17914  42+2  42+2  42+2  39+2  40+2  40+2  40+2  38+2  35+2  35+2  35+2  38+2   
17915  46+2  46+2  46+2  45+2  44+2  44+2  44+2  44+2  38+2  38+2  38+2  44+2   
17916  46+2  46+2  46+2  47+2  45+2  45+2  45+2  46+2  39+2  39+2  39+2  46+2   
17917  44+2  44+2  44+2  45+2  45+2  45+2  45+2  46+2  45+2  45+2  45+2  46+2   

        LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0      64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1      65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2      65+3  60+3  60+3  60+3  65+

Usarem la mateixa tècnica per transformar els atributs "Nationality", "Work Rate", "Position", i "Preferred Foot" ja que tenen la mateixa problemàtica.

In [18]:
nat = df.pop("Nationality")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nat, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

wr = df.pop("Work Rate")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(wr, prefix='wr').reset_index(drop=True)], axis=1, sort=False)

pos  = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

pf  = df.pop("Preferred Foot")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pf, prefix='pf').reset_index(drop=True)], axis=1, sort=False)

df.head()

Age  Overall  Potential    Value   Wage  International Reputation  \
0   31       94         94  €110.5M  €565K                       5.0   
1   33       94         94     €77M  €405K                       5.0   
2   26       92         93  €118.5M  €290K                       5.0   
3   27       91         93     €72M  €260K                       4.0   
4   27       91         92    €102M  €355K                       4.0   

   Weak Foot  Skill Moves Contract Valid Until Height  Weight    LS    ST  \
0        4.0          4.0                 2021    5'7  159lbs  88+2  88+2   
1        4.0          5.0                 2022    6'2  183lbs  91+3  91+3   
2        5.0          5.0                 2022    5'9  150lbs  84+3  84+3   
3        3.0          1.0                 2020    6'4  168lbs   0+0   0+0   
4        5.0          4.0                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3       68.0     15.0            21.0           13.0 

### Conversió de valors numèrics expresats en Strings a numèrics

En aquest dataset existeixen molts valors numèrics expresats en forma d'String pugui ser per haver afegit alguna mena de prefixe o sufixe en les dades, com volem dades purament numèriques, aquest prefixe o sufixe s'haurà d'eliminar.

Per transformar els valor monetaris com "Value", "Wage" i "Release Clause" a un nombre enter farem el següent

In [19]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

In [20]:
df["Value"] = df["Value"].apply(value_to_float)
df["Wage"] = df["Wage"].apply(value_to_float)
df["Release Clause"] = df["Release Clause"].apply(value_to_float)
df.head()


Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         93   72000000.0  260000.0                       4.0   
4   27       91         92  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves Contract Valid Until Height  Weight    LS    ST  \
0        4.0          4.0                 2021    5'7  159lbs  88+2  88+2   
1        4.0          5.0                 2022    6'2  183lbs  91+3  91+3   
2        5.0          5.0                 2022    5'9  150lbs  84+3  84+3   
3        3.0          1.0                 2020    6'4  168lbs   0+0   0+0   
4        5.0          4.0                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3       68.

M'he adonat que "Contract Valid Until" té les dades en dos formats diferents, únicament mostrant l'any de la validesa del contracte o especificant també el dia i el mes. Convertirem tots els registres al primer format: 

In [21]:
df['Contract Valid Until'].unique()

array(['2021', '2022', '2020', '2023', '2019', '2024', 'Jun 30, 2019',
       '2025', '2026', 'Dec 31, 2018', '2018', 'May 31, 2020',
       'Jun 30, 2020', 'May 31, 2019', 'Dec 31, 2019', 'Jan 1, 2019',
       'Jun 1, 2019', 'Jan 4, 2019', 'Jan 31, 2019', 'Jan 7, 2019',
       'Jan 2, 2019', 'Jan 6, 2019', 'Oct 14, 2019', 'Jan 3, 2019',
       'May 4, 2019', 'Jan 12, 2019', 'Jan 25, 2019', 'Jan 18, 2019',
       'Dec 1, 2019', 'Nov 30, 2018', 'Feb 27, 2020', 'Jan 5, 2019',
       'Jan 15, 2019', 'Jan 30, 2019', 'Jan 11, 2019', 'Jan 20, 2019'],
      dtype=object)

In [22]:
def date_to_year(x):
    """
    From date (DD/MM/YYYY) to YYYY.
    
    """
    # Si la es troba en format YYYY directament es fa el return
    if(len(x) == 4):
        return x
    year = x[-4:]
    return year

In [23]:
df["Contract Valid Until"] = df["Contract Valid Until"].apply(date_to_year)
df['Contract Valid Until'].unique()

array(['2021', '2022', '2020', '2023', '2019', '2024', '2025', '2026',
       '2018'], dtype=object)

L'altura dels jugadors "Height", està expresada en peus (X'Y), per tant es passarà a centímetres: 

In [24]:
def to_cm(x):
    """
    From USA height format to SI.
    
    """
    height = x.split('\'')
    feet = int(height[0])
    inches = int(height[1])
    cm = feet*30.48 + inches*2.54
    return cm

In [25]:
df["Height"] = df["Height"].apply(to_cm)
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         93   72000000.0  260000.0                       4.0   
4   27       91         92  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves Contract Valid Until  Height  Weight    LS    ST  \
0        4.0          4.0                 2021  170.18  159lbs  88+2  88+2   
1        4.0          5.0                 2022  187.96  183lbs  91+3  91+3   
2        5.0          5.0                 2022  175.26  150lbs  84+3  84+3   
3        3.0          1.0                 2020  193.04  168lbs   0+0   0+0   
4        5.0          4.0                 2023  180.34  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3    

Finalment eliminarem les unitats de mesura en el pes dels jugadors "Weight", expresat en lbs

In [26]:
def weight_to_int(x):
    """
    Remove lbs at the end of every "Weight" value
    
    """
    return x[:len(x)-3]

In [27]:
df["Weight"] = df["Weight"].apply(weight_to_int)
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         93   72000000.0  260000.0                       4.0   
4   27       91         92  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves Contract Valid Until  Height Weight    LS    ST  \
0        4.0          4.0                 2021  170.18    159  88+2  88+2   
1        4.0          5.0                 2022  187.96    183  91+3  91+3   
2        5.0          5.0                 2022  175.26    150  84+3  84+3   
3        3.0          1.0                 2020  193.04    168   0+0   0+0   
4        5.0          4.0                 2023  180.34    154  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   0+0   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      17.0       13.0             21.0          50.0     13.0       18.0   
4      93.0       82.0             55.0          92.0     82.0       86.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   21.0        19.0         51.0         42.0          57.0         58.0   
4   85.0        83.0         91.0         91.0          78.0         76.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     60.0       90.0     43.0       31.0     67.0     43.0      64.0   
4     79.0       91.0     77.0       91.0     63.0     90.0      75.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       12.0        38.0           30.0         12.0    68.0       40.0   
4       91.0        76.0           61.0         87.0    94.0       79.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3       68.

Per acabar, els valors representats en *LS, ST, RS, LW, LF, CF, RF, RW, LAM, CAM, RAM, LM, LCM, CM, RCM, RM, LWB, LDM, CDM, RDM, RWB, LB, LCB, CB, RCB i RB* estan amb el format: "Nombre base + extra", aquest extra determinat per la composició de la plantilla, com es interesen mirar als jugadors individualment, es deixarà unicament el nombre base. 

In [28]:
def to_base(x):
    t = x.find("+")
    return x[:t]

In [29]:
df["LS"] = df["LS"].apply(to_base)
df["ST"] = df["ST"].apply(to_base)
df["RS"] = df["RS"].apply(to_base)
df["LW"] = df["LW"].apply(to_base)
df["LF"] = df["LF"].apply(to_base)
df["CF"] = df["CF"].apply(to_base)
df["RF"] = df["RF"].apply(to_base)
df["RW"] = df["RW"].apply(to_base)
df["LAM"] = df["LAM"].apply(to_base)
df["CAM"] = df["CAM"].apply(to_base)
df["RAM"] = df["RAM"].apply(to_base)
df["LM"] = df["LM"].apply(to_base)
df["LCM"] = df["LCM"].apply(to_base)
df["CM"] = df["CM"].apply(to_base)
df["RCM"] = df["RCM"].apply(to_base)
df["RM"] = df["RM"].apply(to_base)
df["LWB"] = df["LWB"].apply(to_base)
df["LDM"] = df["LDM"].apply(to_base)
df["CDM"] = df["CDM"].apply(to_base)
df["RDM"] = df["RDM"].apply(to_base)
df["RWB"] = df["RWB"].apply(to_base)
df["LB"] = df["LB"].apply(to_base)
df["LCB"] = df["LCB"].apply(to_base)
df["CB"] = df["CB"].apply(to_base)
df["RCB"] = df["RCB"].apply(to_base)
df["RB"] = df["RB"].apply(to_base)

df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         93   72000000.0  260000.0                       4.0   
4   27       91         92  102000000.0  355000.0                       4.0   

   Weak Foot  Skill Moves Contract Valid Until  Height Weight  LS  ST  RS  LW  \
0        4.0          4.0                 2021  170.18    159  88  88  88  92   
1        4.0          5.0                 2022  187.96    183  91  91  91  89   
2        5.0          5.0                 2022  175.26    150  84  84  84  89   
3        3.0          1.0                 2020  193.04    168   0   0   0   0   
4        5.0          4.0                 2023  180.34    154  82  82  82  87   

   LF  CF  RF  RW LAM CAM RAM  LM LCM  CM RCM  RM LWB LDM CDM RDM RWB  LB LCB  \
0  93  93  93  92  93  93  93  91  84  84  84  91  64  61  61  61  64  59  47   
1  90  90  90  89  88  88  88  88  81  81  81  88  65  61  61  61  65  61  53   
2  89  89  89  89  89  89  89  88  81  81  81  88  65  60  60  60  65  60  47   
3   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   
4  87  87  87  87  88  88  88  88  87  87  87  88  77  77  77  77  77  73  66   

   CB RCB  RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47  47  59      84.0       95.0             70.0          90.0     86.0   
1  53  53  61      84.0       94.0             89.0          81.0     87.0   
2  47  47  60      79.0       87.0             62.0          84.0     84.0   
3   0   0   0      17.0       13.0             21.0          50.0     13.0   
4  66  66  73      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      15.0   

   GKHandling  GKKicking  GKPositioning  GKReflexes  Release Clause  \
0        11.0       15.0           14.0         8.0     226500000.0   
1        11.0       15.0           14.0        11.0     127100000.0   
2         9.0       15.0           15.0        11.0     

## Predicció

Una vegada s'ha transformat el dataset a un format apte per treballar, procedirem a realitzar la predicció. Per això usarem la llibreria sklearn, collint un conjunt del dataset aleatori, establint una seed perque sempre esculli els mateixos registres.

In [30]:
val = df.pop("Value")
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=42)

In [31]:
len(X_train)

12005

In [32]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

In [33]:
reg

LinearRegression()

In [34]:
preds = reg.predict(X_test)

In [35]:
preds[0]

-370107.52766278386

In [36]:
y_test[0]

110500000.0

In [37]:
r2_score(preds, y_test)

0.9666033463039603

Veim que obtenim un RScore del 96.66%, un resultat acceptable. Aquí finalitza aquesta tercera pràctica de l'assignatura 21722 - Intel·ligència Artificial.